### PR TITLE
83447690 better block deleting

### PR DESCRIPTION
**Pull request by islemaster**
Manually pulling some block-deletion changes from [upstream blockly](https://developers.google.com/blockly/).
- Adds a delete-hand cursor to help indicate when a block is being held over a delete zone.
- Cleans up handling of delete zones somewhat.
- Makes toolbox (with categories) a delete zone, to match the behavior of our non-category flyouts and mutator interfaces.
- Applies consistent bump logic to toolbox+category levels to match non-category flyout levels, and help prevent loss of undeleteable blocks by dropping them beneath the toolbox.
- Removes the unused concept of 'concrete blocks' since they were just complicating the delete logic.

Also satisfies part of [#57694140](https://www.pivotaltracker.com/n/projects/914728/stories/57694140).

### CODE REVIEW

File: blockly-core/core/block_space_editor.js [hunk 1]
```diff
@@ -233,8 +238,25 @@ Blockly.BlockSpaceEditor.prototype.addFlyout_ = function() {
...
+
+  if (this.flyout_) {
+    goog.array.extend(deleteAreas, this.flyout_.getRect());
+  }
+
+  if (this.toolbox) {
+    goog.array.extend(deleteAreas, this.toolbox.getRect());
+  }
 
-  this.addChangeListener(this.flyoutBumpOrDeleteOutOfBoundsBlocks_);
```
Comment (islemaster) [inline blockly-core/core/block_space_editor.js]: Renamed method and moved hookup to createDom_ in this file, as the bump logic is no longer dependent on a flyout being present.

File: blockly-core/core/block_space_editor.js [hunk 2]
```diff
@@ -266,13 +288,6 @@ Blockly.BlockSpaceEditor.prototype.flyoutBumpOrDeleteOutOfBoundsBlocks_ = functi
   for (var b = 0, block; block = blocks[b]; b++) {
     var blockXY = block.getRelativeToSurfaceXY();
     var blockHW = block.getHeightWidth();
-    // Have flyout handle any blocks that have been dropped on it
-    if (block.isDeletable() && (Blockly.RTL ?
-        blockXY.x - 2 * metrics.viewLeft - metrics.viewWidth :
-        -blockXY.x) > MARGIN * 2) {
-      this.flyout_.onBlockDropped(block);
-      return;
-    }
```
Comment (islemaster) [inline blockly-core/core/block_space_editor.js]: This delete is a combined side-effect of removing the concreteBlocks setting and unifying block deletion logic; the flyout no longer has a need to handle blocks dropped onto it (and if it ever does again, that message should probably be hooked up elsewhere).

**Comment by islemaster** [comment]
Pushed another merge from development, capturing a recent staging->development merge and hoping to kick off the Travis build again.

**Comment by islemaster** [comment]
Hmm... I'm not sure what's causing the failed Travis build.  I'm surprised to see it's unable to find rake:

```
$ bin/rake db:create db:schema:load
Could not find rake-10.3.1 in any of the sources
Run `bundle install` to install missing gems.
```

Especially after finding this in Travis' [documentation](http://docs.travis-ci.com/user/languages/ruby/):

> Travis CI runs rake by default to execute your tests. Please note that you need to add rake to your Gemfile (adding it to the :test group should be sufficient).

But then, I'm *very* new to Travis.  I could use another set of eyes on this.

**Comment by laurelfan** [comment]
I fixed travis in staging branch here:

https://github.com/code-dot-org/code-dot-org/commit/ba5d1651aba5787ac15b1c4238b3bf53b54e94fe

https://travis-ci.org/code-dot-org/code-dot-org/builds/46006551

If you merge staging to your branch it should fix your branch.

**Comment by islemaster** [comment]
Unfortunately my branch is based off of development, which hasn't picked up the rename changes from today yet.  I'm not sure I want to merge staging to this feature branch, then try to pull this branch into development; For now I've cherry-picked your change to (hopefully) leave the pull request in a good state tonight, and maybe we can figure out the appropriate series of merges tomorrow.

**Comment by islemaster** [comment]
It sounds like we're trying to move away from using the development branch, so a merge back into development doesn't make much sense at this point.  I'm closing this pull request and I'll open a new one against staging once I make sure it will merge okay.  Thanks!

**Pull request closed without merging**